### PR TITLE
Add support for Postgres triggers with arguments

### PR DESCRIPTION
**Pull request by zimtkeks**
Hello,

first of all, I really appreciate this library and the work that went into it. It's a really useful addition when using an ORM and alembic.

There's one feature that is not supported, but relevant for my use-case, namely triggers that have arguments. Triggers _without_ arguments are supported, though.

Being encouraged by your contribution guideline, I decided to implement changes, that extend `sqlalchemy_declarative_extensions.dialects.postgresql.trigger.Trigger` and related code such that it works for _my_ use-case. With some good will the present state may be considered as a PoC.

Given that there is interest on maintainer-side for adding this feature, I'd like to extend this PR such that it is worthy of being merged (including test cases, documentation, etc.).

I'd like to contribute, but I also need guidance to do so. I'm fairly new to SQLA, postgres DBs and even contributing to an OSS project. I'm not sure how to get some feedback. So for now I hope a draft PR is fine for collecting and discussing code changes.

Please let me know if you're interested in a collaboration (also if you're not). I'll try to keep the workload low on your end. I'll probably only need a few hints. Maybe you have some already :-) (e.g. implementation issue I overlooked, some test that could serve as base to test the feature)

Thank you very much!

### CODE REVIEW

File: src/sqlalchemy_declarative_extensions/dialects/postgresql/trigger.py [hunk 1]
```diff
@@ -87,6 +87,7 @@ class Trigger(base.Trigger):
     time: TriggerTimes
     for_each: TriggerForEach = TriggerForEach.statement
     condition: str | None = None
+    arguments: str | None = None
```
Comment (DanCardin) [inline src/sqlalchemy_declarative_extensions/dialects/postgresql/trigger.py:90]: should this be

```suggestion
    arguments: list[str] | None = None
```

? The docs for [create trigger](https://www.postgresql.org/docs/current/sql-createtrigger.html) imply it'll be a comma delimited list of arguments.

Although i do see that `tgargs` is a byta, which i assume means there'll be some parse/mapping required to get it to/from the comma delimited list in sql syntax

**Comment by DanCardin** [comment]
For sure, i'm interested in (at a minimum) any changeset that increases the overall coverage of the database object we already support!

For most preexisting object types, it should be unnecessary to write any actual alembic test, so the easiest way to go about testing is generally just by copying any [trigger existing test](https://github.com/DanCardin/sqlalchemy-declarative-extensions/blob/main/tests/trigger/test_wrapped_parens.py) and just adapting to cover your usecase.

For this particular feature, i wouldn't think you'd need to worry about specific updating tests because any actual change to the arguments list will realistically require a replacement of the whole trigger. So i figure a trigger with a single argument, and a trigger with 2 arguments probably covers all the meaningful tests you could write.

As for documentation, i think just including a bit about how to use it in the docstring, as well as revising some of the notes in https://github.com/DanCardin/sqlalchemy-declarative-extensions/blob/main/docs/source/dialects/postgresql.md about how it only supports the audit feature is probably the sum total of what you'd need.

Finally, feel free to bump the minor version in the pyproject.toml, and include a note in the changelog. otherwise i'll do so after this gets merged anyway.

File: src/sqlalchemy_declarative_extensions/dialects/postgresql/trigger.py (hunk 1 above)
Comment (zimtkeks) [reply src/sqlalchemy_declarative_extensions/dialects/postgresql/trigger.py]: You're right, I misunderstood that first. Fixed in today's commit.

**Comment by zimtkeks** [comment]
Thanks for your helpful reply!

I revised the parsing of trigger arguments from `pg_trigger.tgargs` (and actually tested it this time..).
While the byte representation is an implementation detail and a bit weird, it [doesn't look like it'll change anytime soon](https://postgrespro.com/list/thread-id/2682531). And using `tgargs` (instead of parsing the output of `pg_get_triggerdef()`) better fits with existing code.

Next, I'll try to come up with test for single/mult-argument triggers.

BTW, is it possible for me to run the test.yml workflow on this PR, just for getting feedback if something is broken?

Thanks!

**Comment by DanCardin** [comment]
Happy to let your PR run CI, but generally i'd expect you to only really require it for bulk testing cross-python-version stuff, which your pr doesn't really deal with. If you're passing `pytest tests/trigger` and `make lint` locally, i suspect you'll have a good time generally.

**Comment by zimtkeks** [comment]
As you suggested, I added a test in form of `test_arguments.py` based on `test_wrapped_parens.py` and made minor changes to project files and documentation. That's all I suppose was necessary.
Please let me know if you spot something I missed.

File: docs/source/dialects/postgresql.md [hunk 2]
```diff
@@ -60,7 +60,8 @@ triggers = Triggers().are(
      Trigger.after("insert", on="foo", execute="gimme")
      .named("on_insert_foo")
      .when("pg_trigger_depth() < 1")
-     .for_each_row(),
+     .for_each_row()
+     .with_arguments(["arg1", "arg2"]),
```
Comment (zimtkeks) [inline docs/source/dialects/postgresql.md:64]: Let's just imagine `gimme` consumes arguments for the sake of the example ;-)

File: tests/trigger/test_arguments.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+    functions = Functions().are(
+        Function(
+            "count_arguments",
+            """
+            DECLARE
+                arg_name TEXT;
+            BEGIN
+                FOREACH arg_name in ARRAY TG_ARGV LOOP
+                    RAISE NOTICE 'Processing %', arg_name;
```
Comment (zimtkeks) [inline tests/trigger/test_arguments.py:30]: The notice is pointless but I found it helpful during manual testing on the DB. I tried to consume the emitted message programmatically, but I didn't find a way to do so. Therefore I added the `arg_count` column.

File: tests/trigger/test_arguments.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        )
+    )
+    triggers = Triggers().are(
+        Trigger.after("insert", on="foo", execute="count_arguments")
+        .named("on_insert_foo_single")
+        .for_each_row()
+        .with_arguments(["single"]),
+        Trigger.after("insert", on="foo", execute="count_arguments")
+        .named("on_insert_foo_multi")
+        .for_each_row()
```
Comment (zimtkeks) [inline tests/trigger/test_arguments.py:47]: I tried to define the triggers without `for_each_row()` but then `NEW` was no longer defined in the function.

File: src/sqlalchemy_declarative_extensions/dialects/postgresql/trigger.py [hunk 5]
```diff
@@ -133,6 +134,9 @@ def for_each_statement(self):
     def when(self, condition: str):
         return replace(self, condition=condition)
 
+    def with_arguments(self, arguments: list[str]):
```
Comment (DanCardin) [inline src/sqlalchemy_declarative_extensions/dialects/postgresql/trigger.py:137]: For a builder method like this, what do you think of:

```suggestion
    def with_arguments(self, *arguments: str):
```

two fewer characters at the call site

**Review by DanCardin — APPROVED** [review]

**Comment by zimtkeks** [comment]
Hm, is there anything I can do to fix the failures? They seem unrelated to this PR..

**Comment by DanCardin** [comment]
unfortunately, i've noticed some relatively rare CI-only test randomness because of the network/port stuff going on to make these tests run in parallel.

**Comment by zimtkeks** [comment]
Thank you for supporting and merging my first OSS contribution :)

I'm looking forward to the 0.14 release. Do you have an idea when that might be possible? Thanks!

**Comment by DanCardin** [comment]
Sorry about that, should be released now!

> Thank you for supporting and merging my first OSS contribution :)
No problem at all, thanks for the contribution!

**Pull request merged**